core/src/selection/selectionQuery.h

```diff
@@ -0,0 +1,43 @@
+#pragma once
+
+#include "glm/vec2.hpp"
+#include "tangram.h"
+#include "util/variant.h"
+#include <map>
+
+namespace Tangram {
+
+class MarkerManager;
+class FrameBuffer;
+class TileManager;
+class Labels;
+class View;
+
+enum class QueryType {
+    feature,
+    marker,
+    label,
+};
+
+using QueryCallback = variant<FeaturePickCallback, LabelPickCallback, MarkerPickCallback>;
+
+class SelectionQuery {
+
+public:
+    SelectionQuery(glm::vec2 _position, QueryCallback _queryCallback, QueryType _type);
+
+    static void clearColorCache() { s_colorCache.clear(); }
+
+    void process(const View& _view, const FrameBuffer& _framebuffer, const MarkerManager& _markerManager,
+                 const TileManager& _tileManager, const Labels& _labels) const;
+
+    QueryType type() const { return m_type; }
+
+private:
+    glm::vec2 m_position;
+    QueryCallback m_queryCallback;
+    QueryType m_type;
+
+    static std::map<size_t, uint32_t> s_colorCache;
```

Comment:
I see, then it is probably prudent to cache the results. 

It seems like a poor data structure because in all of our current use cases the map will be storing exactly one entry. I would probably use a vector or forward list instead and do a linear search over the entries. 

As for making the cache non-static, we can allow the call site to create a cache object on the stack and pass it into the processing calls as it loops over the queries.